### PR TITLE
Add Bert flash attention2

### PR DESCRIPTION
**Pull request by sorenmc**
# What does this PR do?

This introduces flash attention 2 for bert as discussed in  https://github.com/huggingface/transformers/issues/26350

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

File: src/transformers/models/bert/modeling_bert.py
Comment (Chiyasa) [inline src/transformers/models/bert/modeling_bert.py:1]: line 795-797 might have problems, because after self.get_extended_attention_mask, attention mask is 0 in position should be attended to, and very large negative number in position should not be attended to. NOT 1 in position should not be attended to?

Comment (Chiyasa) [inline src/transformers/models/bert/modeling_bert.py:1]: line 795-797 might have problems? because in transformers, after get_extended_attention_mask, attention mask is 0 where positions attended to, very negative numbers where positions not attended to. Not 1 where positions not attended to?

**Comment by ArthurZucker** [comment]
cc @younesbelkada

**Review by younesbelkada — COMMENTED** [review]
Hi @sorenmc 
Thanks for the PR, let me know once this is ready for review !

**Comment by sorenmc** [comment]
Hi @younesbelkada 

I finished most of the work for bert, but as you can see `check_repository_consistency` fails because several models are linked to BERT, meaning that i would have to do this same implementation for 10+ models. Might need some help with that, or figure out an alternative way to split some of these up into smaller tasks that could be included in a later PR. Also tests are failing for `Wav2Vec2` that i have not touched - I'm guessing this could be related to the consistency checks?

File: src/transformers/models/bert/modeling_bert.py
Comment (sorenmc) [reply src/transformers/models/bert/modeling_bert.py:1]: Great catch - didn't see that the last line multiplied with the lowest possible value for the dtype (extract from `get_extended_attention_mask` as can be seen below). Will definitely fix this.
```python
# Since attention_mask is 1.0 for positions we want to attend and 0.0 for
# masked positions, this operation will create a tensor which is 0.0 for
# positions we want to attend and the dtype's smallest value for masked positions.
# Since we are adding it to the raw scores before the softmax, this is
# effectively the same as removing these entirely.
extended_attention_mask = extended_attention_mask.to(dtype=dtype)  # fp16 compatibility
extended_attention_mask = (1.0 - extended_attention_mask) * torch.finfo(dtype).min
return extended_attention_mask
```

**Comment by ArthurZucker** [comment]
You should first rebase on main, then run `make fix-copies` that changes will be automatically ported to architectures that are similar to bert. But yes, this requires adding tests using `Copied from` as well in the testing files of all affected models!

**Comment by ArthurZucker** [comment]
Unrelated test are fixed in main 🤗
